### ISSUE TITLE
Ruby Omnicompletion function locks UI 

### ISSUE THREAD

**Issue by stephenprater** [neovim/neovim#5204]
- `nvim --version`:

NVIM v0.1.5-721-gabb0928

- Vim (version: ) behaves differently?

Yes... but it doesn't really work either.

- Operating system/version:

OSX 10.11

- Terminal name/version:

Iterm 2.0
- `$TERM`:

screen-256color

### Actual behaviour

UI locks for several seconds in ruby omni-complete function is invoked.

### Expected behaviour

UI should not lock

### Steps to reproduce using `nvim -u NORC`

Open a file with a ruby filetype.  Press C-x-C-o to invoke the omnifunc.

I believe this is a function of the way "language" plugins work now.  In the before now, a ruby process ran and you cold load the omni-completion function into the persistent interpreter with execute.  It doesn't really work that way anymore and running "ruby execute" with a big long string will execute the string individually each time, not affecting global state.  Better in most regards, but the shipping omni-completion function is pretty well broken since it depends on the old behavior.

### Workaround ###

If you are using deoplete, remove the included omni-complete function from the sources like so:

```
let g:deoplete#ignore_sources = {}
let g:deoplete#ignore_sources.ruby = ['omni']
```

If you are not using deoplete - change the line at the beginning of the `rubycomplete.vim` function in the share/autoload directory from `if !has('ruby')` to `if has('ruby')`

### Actual Fix ###

Rewrite the ruby omni-complete function to be compatible with the new `if_ruby` scheme.

**Comment by alexgenco** [comment]
I can't reproduce... I get the omnicomplete menu, same as `vim`. You're getting this even with `-u NORC`?

**Comment by stephenprater** [comment]
Yes - I'm in a largish Rails app, and it locks it for 2-3 seconds everytime I hit invoke the completion function.  It does seem to work on single files and it does eventually return results.

I'm using the bog-standard vim-ruby configuration.  With regular vim I get "pattern not found" - so perhaps it's a problem with the vim-ruby configuration out of the box.  It also seems to lock at random intervals.

**Comment by Tranquility** [comment]
It works for me. Maybe you installed some plugin that overwrote the completion function?

**Comment by Shougo** [comment]
It seems just slow vim-ruby plugin problem.

**Comment by stephenprater** [comment]
@Shougo I agree - I don't think this is a "bug" in Neovim - I think it's because the provided omni-complete function and the new ruby provider don't play together as well as they should.   (FTR:  using it with deoplete locks it up _entirely_.) 

What's the solution, do you think?  Rewrite the omni-complete function?  Just mark it as incompatible with Neovim and have a plugin?

**Comment by Shougo** [comment]
> What's the solution, do you think? Rewrite the omni-complete function? Just mark it as incompatible with Neovim and have a plugin?

I don't think vim-ruby implementation is good.
I think you should create new Ruby omnifunc.
For example, deoplete source for Ruby.

**Comment by alexgenco** [comment]
@stephenprater if this is performing significantly worse against the neovim ruby provider, we should probably treat it as a regression and see where the bottleneck was introduced. I was thinking you could wrap [this line](https://github.com/vim-ruby/vim-ruby/blob/master/autoload/rubycomplete.vim#L210) in a [`RubyProf.profile` block](https://github.com/ruby-prof/ruby-prof#usage) and see where neovim-ruby is being slow.

**Comment by ianks** [comment]
Same thing happens with me; and I seem to be getting random seg faults as well as a result.

**Comment by justinmk** [comment]
@ianks If you can get a backtrace for the segfault please open a new issue if there isn't an existing one that covers it.
